### PR TITLE
docs: Clarify the coala in a container docs

### PR DESCRIPTION
**Pull request by acashmkj**
Added explanation for Docker.

Closes https://github.com/coala/documentation/issues/30

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by acashmkj** [comment]
@IpshitaC please review

File: Users/Docker_Image.rst [hunk 1]
```diff
@@ -1,3 +1,44 @@
+What is Docker ?
+================
+
+Docker is a tool designed to make it easier to create, deploy, and run
+applications by using containers. Containers allow a developer to package
+up an application with all of the parts it needs, like libraries and other
+dependencies, and ship it all out as a single package.
+
+With the help of containers, the developer can rest assured that
+the application will run on other machine regardless of any
```
Comment (Makman2) [inline Users/Docker_Image.rst]: either
1. on **an**other machine
or
2. on other machine**s**

Comment (Makman2) [reply Users/Docker_Image.rst]: Would prefer 2nd :+1:

File: Users/Docker_Image.rst [hunk 2]
```diff
@@ -1,3 +1,44 @@
...
+customized settings that machine might have that could differ from the
+machine used for writing and testing the code.
+
+Unlike a virtual machine, rather than creating a whole virtual operating
+system, Docker allows applications to use the same kernel as the system
+that they're running on and only requires applications be shipped
```
Comment (Makman2) [inline Users/Docker_Image.rst]: be --> being

File: Users/Docker_Image.rst [hunk 3]
```diff
@@ -1,3 +1,44 @@
...
+with things not already running on the host computer. This gives a
+significant performance boost and reduces the size of the application.
+
+Whats is a Docker Image and how is it different from a container ?
+------------------------------------------------------------------
+
+Image is an immutable file and a running instance of an image is called a
```
Comment (Makman2) [inline Users/Docker_Image.rst]: --> An image

File: Users/Docker_Image.rst [hunk 4]
```diff
@@ -1,3 +1,44 @@
...
+customized settings that machine might have that could differ from the
+machine used for writing and testing the code.
+
+Unlike a virtual machine, rather than creating a whole virtual operating
+system, Docker allows applications to use the same kernel as the system
+that they're running on and only requires applications be shipped
+with things not already running on the host computer. This gives a
+significant performance boost and reduces the size of the application.
+
+Whats is a Docker Image and how is it different from a container ?
```
Comment (Makman2) [inline Users/Docker_Image.rst]: --> What~~s~~

Comment (Makman2) [inline Users/Docker_Image.rst]: Redundant space before question mark

Comment (Makman2) [inline Users/Docker_Image.rst]: ~~as "image" is quite a general purpose word, I would make it lower-case completely^^~~

File: Users/Docker_Image.rst [hunk 5]
```diff
@@ -1,3 +1,44 @@
...
+------------------------------------------------------------------
+
+Image is an immutable file and a running instance of an image is called a
+container. Images are composed of layers of other images. Images are created
+when we run ``build`` command and conatiners are formed from these images
```
Comment (Makman2) [inline Users/Docker_Image.rst]: **the** ``build`` command **of Docker**
(Imo it's important to see where the command comes from, not from the OS, but docker ;D)

Comment (Makman2) [inline Users/Docker_Image.rst]: Typo:
conatiners --> containers

File: Users/Docker_Image.rst [hunk 6]
```diff
@@ -1,3 +1,44 @@
...
+when we use ``run`` command. There can be many containers for the
+same image.
```
Comment (Makman2) [inline Users/Docker_Image.rst]: I think that doesn't really explain what an image is, currently it's some abstract file that consists of layers and can magically be formed into a container.
We should explain that the image contains the data as applications, files, etc to form our container environment (I hope that's actually correct...)

File: Users/Docker_Image.rst [hunk 7]
```diff
@@ -1,3 +1,44 @@
...
+
+  For more imformation about Docker see the
+  `official documentation <https://docs.docker.com/>`__.
```
Comment (Makman2) [inline Users/Docker_Image.rst]: Redundant leading spaces

Comment (Makman2) [inline Users/Docker_Image.rst]: Typo:
I~~m~~nformation

File: Users/Docker_Image.rst [hunk 8]
```diff
@@ -1,3 +1,44 @@
...
+
+Installing Docker
+-----------------
+
+  Docker installation guide for various operating systems can be found
+  `here <https://docs.docker.com/engine/installation/>`__.
```
Comment (Makman2) [inline Users/Docker_Image.rst]: Optional:
I'm not a fan of using `here`-links, rather describe what's linked, like
```
... can be found in the `official Docker installation instructions <...>`__.
```

But like said, optional and up to you ;)

File: Users/Docker_Image.rst [hunk 9]
```diff
@@ -1,3 +1,44 @@
...
+
+Downloading Docker images and pushing images over low bandwidth connections
+is very slow.
```
Comment (Makman2) [inline Users/Docker_Image.rst]: Is this relevant? :)

File: Users/Docker_Image.rst [hunk 10]
```diff
@@ -1,3 +1,44 @@
...
+With the help of containers, the developer can rest assured that
+the application will run on other machine regardless of any
+customized settings that machine might have that could differ from the
+machine used for writing and testing the code.
+
+Unlike a virtual machine, rather than creating a whole virtual operating
+system, Docker allows applications to use the same kernel as the system
+that they're running on and only requires applications be shipped
+with things not already running on the host computer. This gives a
+significant performance boost and reduces the size of the application.
```
Comment (Makman2) [inline Users/Docker_Image.rst]: which application? Applications inside a container don't take less size :O (not sure if docker uses compression, however)

File: Users/Docker_Image.rst [hunk 11]
```diff
@@ -1,3 +1,44 @@
...
+================
+
+Docker is a tool designed to make it easier to create, deploy, and run
+applications by using containers. Containers allow a developer to package
+up an application with all of the parts it needs, like libraries and other
+dependencies, and ship it all out as a single package.
+
+With the help of containers, the developer can rest assured that
+the application will run on other machine regardless of any
+customized settings that machine might have that could differ from the
```
Comment (Makman2) [inline Users/Docker_Image.rst]: maybe "that" --> "his" (the machine of the developer)

File: Users/Docker_Image.rst [hunk 12]
```diff
@@ -1,3 +1,44 @@
...
+with things not already running on the host computer. This gives a
+significant performance boost and reduces the size of the application.
+
+Whats is a Docker Image and how is it different from a container ?
+------------------------------------------------------------------
+
+Image is an immutable file and a running instance of an image is called a
+container. Images are composed of layers of other images. Images are created
+when we run ``build`` command and conatiners are formed from these images
+when we use ``run`` command. There can be many containers for the
```
Comment (Makman2) [inline Users/Docker_Image.rst]: **the** ``run`` command

File: Users/Docker_Image.rst (hunk 4 above)
Comment (Makman2) [reply Users/Docker_Image.rst]: nvm, title-case is okay :+1:

File: Users/Docker_Image.rst (hunk 9 above)
Comment (Makman2) [reply Users/Docker_Image.rst]: Ah okay mentioned inside the issue.
Maybe put this inside a note, probably inside the `coala as a Docker Image` section.

File: Users/Docker_Image.rst (hunk 8 above)
Comment (Makman2) [inline Users/Docker_Image.rst]: Redundant leading spaces

File: Users/Docker_Image.rst [hunk 13]
```diff
@@ -1,3 +1,44 @@
...
+same image.
+
+  For more imformation about Docker see the
```
Comment (Makman2) [inline Users/Docker_Image.rst]: Again:
i~~m~~nformation

File: Users/Docker_Image.rst [hunk 14]
```diff
@@ -1,3 +1,44 @@
+What is Docker ?
```
Comment (Makman2) [inline Users/Docker_Image.rst]: redundant space before question mark

**Comment by Makman2** [comment]
unack 1bb2d15

File: Users/Docker_Image.rst (hunk 10 above)
Comment (acashmkj) [reply Users/Docker_Image.rst]: @Makman2 yes docker compresses data. The actual amount of data that’s pushed will be compressed before sending,

**Comment by acashmkj** [comment]
@Makman2 @IpshitaC please review

File: Users/Docker_Image.rst [hunk 15]
```diff
@@ -1,3 +1,42 @@
...
+machine used for writing and testing the code.
+
+Unlike a virtual machine, rather than creating a whole virtual operating
+system, Docker allows applications to use the same kernel as the system
+that they're running on and only requires applications being shipped
+with things not already running on the host computer. This gives a
+significant performance boost and reduces the size of the application.
+
+What is a Docker Image and how is it different from a container?
+------------------------------------------------------------------
```
Comment (Makman2) [inline Users/Docker_Image.rst]: dashes not aligned to title :) (two dashes too much)

File: Users/Docker_Image.rst [hunk 16]
```diff
@@ -1,3 +1,42 @@
...
+
+An image is an immutable file which contains required binaries and libraries
+needed to build a container and a running instance of an image is called
```
Comment (Makman2) [inline Users/Docker_Image.rst]: maybe
build --> make
to not confuse it with the `build` command of docker below.

Comment (Makman2) [reply Users/Docker_Image.rst]: or maybe just omit the part:
... needed ~~to build a container~~ and a running ...

File: Users/Docker_Image.rst [hunk 17]
```diff
@@ -33,6 +72,9 @@ on your code with a ``.gitlab-ci.yml``, like this:
   For more information about GitLab CI configuration, consult the
   `official documentation <https://docs.gitlab.com/ce/ci/>`__.
 
+Downloading Docker images and pushing images over low bandwidth connections
+is very slow.
```
Comment (Makman2) [inline Users/Docker_Image.rst]: Okay my bad, it doesn't fit here nicely as well :3

However, I would actually rephrase it, to something like:
```
Docker images are usually very large. Downloading or pushing them over low bandwidth connections can be very slow.
```
What do you say? :)

And maybe we put it inside a `note::` back into ""What is a docker image  and how is it different from a container?"

Comment (acashmkj) [reply Users/Docker_Image.rst]: @Makman2 yes I think we should put it back in What is a docker image section

**Comment by acashmkj** [comment]
@Makman2 @IpshitaC please review

File: Users/Docker_Image.rst [hunk 18]
```diff
@@ -1,3 +1,47 @@
...
+Installing Docker
+-----------------
+
+Docker installation guide for various operating systems can be found in the
+`official Docker installation instructions <https://docs.docker.com/engine/installation/>`__.
+
+.. note::
+
+  Downloading Docker images and pushing images over low bandwidth connections
+  is very slow.
```
Comment (Makman2) [inline Users/Docker_Image.rst]: What about
> Docker images are usually very large. Downloading or pushing them over low bandwidth connections can be very slow.

instead? :)
your choice :+1:

Comment (acashmkj) [reply Users/Docker_Image.rst]: Agree with you. Done necessary changes

**Comment by acashmkj** [comment]
@Makman2 please review

**Comment by Makman2** [comment]
ack 5839e8d

**Comment by Makman2** [comment]
@rultor merge

**Comment by jayvdb** [comment]
See https://github.com/coala/documentation/issues/482

**Pull request merged**